contrib/completion/_fd

```diff
@@ -0,0 +1,210 @@
+
+    + '(no-ignore-full)' # all ignore files
+    '(no-ignore-partial)'{-I,--no-ignore}"[don't respect .(git|fd)ignore and global ignore files]"
+    $no'(no-ignore-partial)*'{-u,--unrestricted}'[alias for --no-ignore, when repeated also alias for --hidden]'
+
+    + no-ignore-partial # some ignore files
+    "(no-ignore-full --no-ignore-vcs)--no-ignore-vcs[don't respect .gitignore files]"
+    "(no-ignore-full --no-global-ignore-file)--no-global-ignore-file[don't respect the global ignore file]"
+
+    + '(case)' # case-sensitivity
+    {-s,--case-sensitive}'[perform a case-sensitve search]'
+    {-i,--ignore-case}'[perform a case-insensitve search]'
+
+    + '(regex-pattern)' # regex-based search pattern
+    '(no-regex-pattern)--regex[perform a regex-based search (default)]'
+
+    + '(no-regex-pattern)' # non-regex-based search pattern
+    {-g,--glob}'[perform a glob-based search]'
+    {-F,--fixed-strings}'[treat pattern as literal string instead of a regex]'
+
+    + '(match-full)' # match against full path
+    {-p,--full-path}'[match the pattern against the full path instead of the basename]'
+
+    + '(follow)' # follow symlinks
+    {-L,--follow}'[follow symbolic links to directories]'
+
+    + '(abs-path)' # show absolute paths
+    '(long-listing)'{-a,--absolute-path}'[show absolute paths instead of relative paths]'
+
+    + '(null-sep)' # use null separator
+    '(long-listing)'{-0,--print0}'[separate search results by the null character]'
+
+    + '(long-listing)' # long-listing output
+    '(abs-path null-sep max-results exec-cmds)'{-l,--list-details}'[use a long listing format with file metadata]'
+
+    + '(max-results)' # max number of results
+    '(long-listing exec-cmds)--max-results=[limit number of search results to given count and quit]:count'
+    '(long-listing exec-cmds)-1[limit to a single search result and quit]'
+
+    + '(error-messages)' # show FS errors
+    $no'--show-errors'
+
+    + '(file-system)' #
+    "--one-file-system[don't descend into directories on other file systems]"
+    '!--mount'
+    '!--xdev'
+
+    + dir-depth # directory depth
+    '(--exact-depth -d --max-depth)'{-d+,--max-depth=}'[set max directory depth to descend when searching]:depth'
+    '!(--exact-depth -d --max-depth)--maxdepth:depth'
+    '(--exact-depth --min-depth)--min-depth=[set directory depth to descend before start searching]:depth'
+    '(--exact-depth -d --max-depth --maxdepth --min-depth)--exact-depth=[only search at the exact given directory depth]:depth'
+
+    + filter-misc # Filter search
+    '*'{-t+,--type=}"[filter search by type]:type:(($fd_file_types))"
+    '*'{-e+,--extension=}'[filter search by file extension]:extension'
+    '*'{-E+,--exclude=}'[exclude files/directories that match the given glob pattern]:pattern'
+    '*'{-S+,--size=}'[limit search by file size]:size limit'
+    '(-o --owner)'{-o+,--owner=}'[filter by owning user and/or group]:owner/group'
```

Comment:
Not sure how this argument works, whether it's supposed to be `user:group` or what, but this could probably call `_users` if nothing else. (The `user:group` thing is so common that we should probably have a helper function for it, but we don't yet. It's not too hard to write out, see e.g. `_chown` or `_chroot`, but it's understandable if you want to avoid that)

Reply:
The arguments for `--owner` can be `user`, `user:group` or `:group`, and then any of those three can be prefixed with `!` to exclude matching files instead (maybe a `--not-owner` option would have been better). I don't understand from the manual if `user!:group` is valid (I don't think so).

I am still trying to figure out more advanced completion with Zsh. I think it can be done, but the optional `!` threw me off.

Reply:
Without having tested it much (some escaping issues with `!`, but this is roughly how other functions do it), maybe something like this:

```
local ret=1
local -a context line state state_descr

# ...
'*'{-o+,--owner=}'[filter by owning user and/or group]: :->owner'
# ...

_arguments ... && ret=0

if [[ $state == owner ]]; then
  if compset -P '*:'; then
    compset -P '(\\|)!'
    _groups && ret=0
  else
    compset -P '(\\|)!'
    _users -qS : && ret=0
  fi
fi

return ret
```

Note the addition of the extra local variables at the top; you need those whenever you have state handling.

Also, i looked at the manual and saw that `--owner` can be supplied more than once, so i added `*` to the arg spec.

Hope that makes sense

Reply:
I wasn't really planning to see how to complete `--owner` and `--size` yet, but you just kinda passed me the ball to do it now 😂 Just joking, your help is much appreciated. I have been trying to learn as much of possible of Zsh completion, I have been writing a few other completion scripts and that's why I ended up doing this one.

I have the `--owner` completion working with this, which is taking the same structure as `_chown` as you mentioned, but I tried to add some fixes due to the optional `!` prefix:

``` zsh
compset -p '(\\|)\!'
if compset -p '*:'; then
  _groups && ret=0
else
  if
    compset -s ':*' ||
    # do not add the colon suffix when completing "!user<tab>
    # (with a starting double-quote) otherwise pressing tab again
    # after the inserted colon "!user:<tab> will complete history modifiers
    [[ $iprefix == (\\|\!)*  && ($qiprefix == \"* && -z $qisuffix) ]]
  then
    _users && ret=0
  else
    local q
    # since quotes are needed when using the negation prefix !,
    # automatically remove the colon suffix also when closing the quote
    if [[ $qiprefix == [\'\"]* ]]; then
      q=${qiprefix:0:1}
    fi
    _users -r ": \t\n\-$q" -s : && ret=0
  fi
fi
```

First, without quotes it works just like `_chown`. But since quotes are needed in order to use the `!` prefix (I tried escaping the `!` but when the completion inserted a full value it would remove the backslash), I tried to fix a couple of annoyances:

- When using double-quotes, the completion will not insert the `:` suffix after the user. This is to fix the next <kbd>Tab</kbd> that is supposed to complete groups but it will show history modifiers instead if no closing double-quote is present. With single-quotes the suffix _is_ inserted. A <kbd>Tab<kbd> after that will complete the group just fine.

- When using quotes and the colon suffix is inserted, I use `-r` instead of `-q` to also enable removing the suffix when typing the closing quote (it just saves the <kbd>Backspace<kbd> to remove the colon and they typing the quote, but I think it is nice to have it)

As a side note, @sharkdp I think it would be better if in the manpage, the `--owner` part shows the exclusion example with single-quotes: `--owner '!john:students'`. The output of `--help` does it.